### PR TITLE
feat: enable rrweb to record and replay log messages in console

### PR DESCRIPTION
**Pull request by Mark-Fenng**
this is the implementation of new feature request(issue #234 )(issue #317 )

here are a few points of description.
1. users need to set recordLog option in rrweb.record's parameter to record log messages.  The log recorder is off by default.
2. support recording and replaying all kinds of console functions. But the reliability of them should be tested more
3. the stringify function in  stringify.ts needs improvement. e.g. robustness and better support for more kinds of object
4. we can replay the log messages in a simulated html console like LogRocket by implementing the interface "ReplayLogger" in the future

### CODE REVIEW

**Comment by Mark-Fenng** [comment]
@Yuyz0112 Could you please review this pull request😂  I've solved merge conflicts two times in the past one week

File: src/types.ts [hunk 1]
```diff
@@ -353,8 +369,84 @@ export type fontParam = {
...
+  table?: (tabularData: any, properties?: ReadonlyArray<string>) => void;
+  time?: (label?: string) => void;
+  timeEnd?: (label?: string) => void;
+  timeLog?: (label?: string, ...data: any[]) => void;
+  trace?: (message?: any, ...optionalParams: any[]) => void;
+  warn?: (message?: any, ...optionalParams: any[]) => void;
+};
+
+// functions to replay log record
+export type ReplayLogger = {
```
Comment (Yuyz0112) [inline src/types.ts:418]:
```ts
export type ReplayLogger = Record<LogLevel, (data: logData) => void>
```

Is this an equivalent?

**Comment by Yuyz0112** [comment]
LGTM

Most of the design looks great and just a few things need to be tweaked.

File: src/types.ts (hunk 1 above)
Comment (Mark-Fenng) [reply src/types.ts:418]: Yes, this one is better!

**Comment by eoghanmurray** [comment]
There is no easy way to transmit events for when an unhandled exception happens on the page?  
I.e. if we don't have a chance to catch and log with `console.error` because the exception happens in 3rd party js code?

**Pull request merged**